<!DOCTYPE html><html lang="en" ><head ><title >Exception handling proof-of-concept</title><meta name='viewport' content='width=device-width, initial-scale=0.66'><meta name='description' content='language:basic, target:desktop, category:game development'><link rel="styleSheet" href="../skins/granite/style.css" type="text/css"><style type="text/css">div.bbcode {padding: 8px;background: #E5E5E5;color: #000000;border: 1px dashed #B4B4BE;}</style></head><body ><table width=100% cellspacing=0 cellpadding=0><tr ><td class="menubarleft"></td><td class="menubar"><table cellspacing="0" cellpadding="0"><tr ><td >&nbsp;</td><td class="menuitemleft"></td><td class="menuitem"><a href="/" class="menuitem">News</a></td><td class="menuitemright"></td><td class="menuitemleft"></td><td class="menuitem"><a href="/forums.php" class="menuitem">Forums</a></td><td class="menuitemright"></td><td class="menuitemleft"></td><td class="menuitem"><a href="/codearcs.php" class="menuitem">Code</a></td><td class="menuitemright"></td><td class="menuitemleft"></td><td class="menuitem"><a href="/logs.php" class="menuitem">Logs</a></td><td class="menuitemright"></td><td class="menuitemleft"></td><td class="menuitem"><a href="/gallery.php" class="menuitem">Gallery</a></td><td class="menuitemright"></td><td class="menuitemleft"></td><td class="menuitem"><a href="/sdkspecs.php" class="menuitem">Specs</a></td><td class="menuitemright"></td><td class="menuitemleft"></td><td class="menuitem"><a href="/search.php" class="menuitem">Search</a></td><td class="menuitemright"></td></tr></table></td><td class="menubarright"></td></tr></table><div class="main"><h1 >Exception handling proof-of-concept</h1><a href="forums.php" >Monkey Forums</a>/<a href="topics.php?forum=507" >Monkey Programming</a>/<a href="#bottom" >Exception handling proof-of-concept</a><br><br>
<a name="2011352"></a>

<a name="2011353"></a>

<a name="2011354"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td class="posthead"><table width=100% cellspacing=0 cellpadding=0><tr ><td >Samah</td><td align="right"><font class=tiny>(Posted 2011)</font>&nbsp;<a href="#1">[#1]</a></td></tr></table></td></tr><tr ><td class="posttext"> I made some modifications to trans just to prove it could be done. Tested HTML5 build in IE8 and FF5.<br><br>Failing example:<br><pre class=code>Function Main:Int()
	Print "trying invalid cast: StringObject(Object(New StringList))"
	Print StringObject(Object(New StringList))
	' program dies with javascript error "TypeError: 'null' is null or not an object"
	' does not print the next line
	Print "continue with the rest of the program"
	Return 0
End</pre><br>Exception handling example:<br><pre class=code>Function Main:Int()
	Try
		Print "trying invalid cast: StringObject(Object(New StringList))"
		Print StringObject(Object(New StringList))
		' program doesn't die from the error and instead jumps to the catch block
	Catch
		Print "caught the error"
	End
	' program continues as if nothing is wrong
	Print "continue with the rest of the program"
	Return 0
End</pre><br>Output:<br><pre class=code>trying invalid cast: StringObject(Object(New StringList))
caught the error
continue with the rest of the program</pre><br>And of course if there was no error, the catch block doesn't get executed.<br><br>This was REALLY easy to add, and all current targets support some form of exception handling.  I've only done the HTML5 translation for now, and the Throw keyword is not implemented.  I also need to make the error message available to the developer in the catch block.<br><br>Given that this is essentially a fork, I don't want to have to maintain it nor merge it with every major release.  Please add this Mark!!! <br><br></td></tr></table><br>
<a name="2011355"></a>

<a name="2011356"></a>

<a name="2011357"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td class="posthead"><table width=100% cellspacing=0 cellpadding=0><tr ><td >marksibly</td><td align="right"><font class=tiny>(Posted 2011)</font>&nbsp;<a href="#2">[#2]</a></td></tr></table></td></tr><tr ><td class="posttext"> Hi,<br><br>It's a bit harder than that - the debug stack state also needs to be tracked, and popped appropriately too.<br><br>If there's a big enough demand, I'll consider adding exceptions, but I do not personally like them and could quite happily live without them.<br><br>To me, the biggest problem is that it makes it much harder for an object to maintain an invariant (ie: a consistent state) if any method call you make can potentially do a 'goto' behind your back. You can never be sure the next instruction will be executed, so any method which performs multiple actions has to be written very carefully.<br><br>Even in your example above, I would prefer having to test an object for null before using it - this at least indicates I know where it can fail. Your code above really has a logic error in it anyway - if a downcast fails, there's something wrong with the code and exceptions can't help fix that...<br><br>Some more (biased!) thoughts on exceptions:<br> <br><a href="http://soundadvice.id.au/blog/2004/10/16/" target="_blank">http://soundadvice.id.au/blog/2004/10/16/</a><br><a href="http://schlitt.info/opensource/blog/0724_python_good_bad_evil_03_flow_control_exceptions.html" target="_blank">http://schlitt.info/opensource/blog/0724_python_good_bad_evil_03_flow_control_exceptions.html</a><br><a href="http://neil.fraser.name/writing/exception/" target="_blank">http://neil.fraser.name/writing/exception/</a><br><a href="http://ptgmedia.pearsoncmg.com/images/020163371x/supplements/Exception_Handling_Article.html" target="_blank">http://ptgmedia.pearsoncmg.com/images/020163371x/supplements/Exception_Handling_Article.html</a><br><br>I don't necessarily agree that exceptions are 'evil', but I do believe they are MUCH harder to use safely than most people realize. <br><br></td></tr></table><br>
<a name="2011364"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td class="posthead"><table width=100% cellspacing=0 cellpadding=0><tr ><td >Samah</td><td align="right"><font class=tiny>(Posted 2011)</font>&nbsp;<a href="#3">[#3]</a></td></tr></table></td></tr><tr ><td class="posttext"> <div class="quote"> You can never be sure the next instruction will be executed, so any method which performs multiple actions has to be written very carefully. <br></div><br>This is the difference between a checked and unchecked exception.  In Java and other languages, checked exceptions must be declared to be thrown.  It's not possible to write code that ignores them.  Unchecked exceptions (array index out of bounds, null pointer, etc.) can be thrown at any time (as they currently are) so are only caught if you explicitly declare that you want to.<br><br><div class="quote"> Your code above really has a logic error in it anyway - if a downcast fails, there's something wrong with the code and exceptions can't help fix that... <br></div><br>I just quickly put something in to make it fail at runtime.  I was going to use an invalid Int cast, but it seems that only fails on Android.<br><br>An example where you might want to use exception handling rather than some arbitrary result:<br><pre class=code>Function Main()
	Local s:String = "foo"
	Print Int(s)
End</pre>This prints "NaN".<br><br><pre class=code>Function Main()
	Print Int("foo")
End</pre>This will print "0" (although the developer would be silly to hardcode this anyway).<br><br><pre class=code>Function Main()
	Print Int(" 1")
End</pre>This will print "1" on all targets (I think) except Android, which will fail unless you Trim() it first.<br><br>If the string you're casting is not a valid integer, I would expect it to throw an unchecked exception.  Otherwise you would need to iterate through every character in the string to ensure it's numeric before you try to cast it, in case you get a result you weren't expecting.<br><br>Something like this would be much better:<br><pre class=code>Function Main()
	Local s:String = "foo"
	Try
		Print Int(s)
	Catch ex:CastException
		Print ex.Message
	End
End</pre>Of course you probably shouldn't be relying on exceptions for that kind of thing, but it's just an example.<br><br><div class="quote"> I don't necessarily agree that exceptions are 'evil', but I do believe they are MUCH harder to use safely than most people realize. <br></div><br>I agree that they can be misused, but if you don't fully understand them, you probably shouldn't be using them.  But you could say that for a lot of paradigms, and it doesn't stop newish programmers (or even experienced ones) from making silly mistakes.<br><br>At least give some thought to it for developers who wish to use them.  If you're really worried about the "evils of random gotos", maybe only allow user-defined exception handling (the Throw and Throws keywords), and any random system generated ones will just explode as they currently do.<br><br><div class="quote"> Some more (biased!) thoughts on exceptions: <br></div><br>Aww, that's a little unfair. :)<br><br>Edit:<br>Pleeaasseee give us a forum post preview button! I've taken to hitting "reply" on a random post in the blitz forums so I can use the preview on there. <br><br></td></tr></table><br>
<a name="2011363"></a>

<a name="2011362"></a>

<a name="2011361"></a>

<a name="2011360"></a>

<a name="2011358"></a>

<a name="2011359"></a>

<a name="2011366"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td class="posthead"><table width=100% cellspacing=0 cellpadding=0><tr ><td >marksibly</td><td align="right"><font class=tiny>(Posted 2011)</font>&nbsp;<a href="#4">[#4]</a></td></tr></table></td></tr><tr ><td class="posttext"> Hi,<br><br>&gt; It's not possible to write code that ignores them.<br><br>Yes it is - you just add 'throws Blah' to the prototype. This way your entire method can just pass on the exception and ignore it totally. I used to do it all the time back at school!<br><br>&gt; This prints "NaN".<br><br>Oops - that's a bug! Looks like the const precompiler acting up again...<br><br>&gt; Of course you probably shouldn't be relying on exceptions for that kind of thing, but it's just an example.<br><br>Yes, good examples of exception use are hard to come by!<br><br>&gt; At least give some thought to it for developers who wish to use them.<br><br>Will do... <br><br></td></tr></table><br>
<a name="2011365"></a>

<a name="2011367"></a>

<a name="2011368"></a>

<a name="2011369"></a>

<a name="2011370"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td class="posthead"><table width=100% cellspacing=0 cellpadding=0><tr ><td >Samah</td><td align="right"><font class=tiny>(Posted 2011)</font>&nbsp;<a href="#5">[#5]</a></td></tr></table></td></tr><tr ><td class="posttext"> <div class="quote"> Yes it is - you just add 'throws Blah' to the prototype. This way your entire method can just pass on the exception and ignore it totally. I used to do it all the time back at school! <br></div><br>That's not ignoring it though, that's deferring it. :)<br><br>My point is that checked exceptions must be caught SOMEWHERE in the call stack, unless you declare every method (including Main) to throw it.  Unchecked exceptions do not need to be caught or declared as thrown - they're implicitly thrown.  If you don't catch it somewhere, it'll simply get bumped up the call stack and crash the thread (as it should).<br><br>Edit:<br>C#/.NET has a cool TryParse method so that you can parse strings to primitives and fail silently with a false return value.  Can we have something like this?  A native implementation would be nice. :)<br><pre class=code>If IsInt(mystring) Then 'do int stuff</pre> <br><br></td></tr></table><br>
<a name="2011371"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td class="posthead"><table width=100% cellspacing=0 cellpadding=0><tr ><td >Roger Lockerbie</td><td align="right"><font class=tiny>(Posted 2011)</font>&nbsp;<a href="#6">[#6]</a></td></tr></table></td></tr><tr ><td class="posttext"> Excuse me if this is really dumb as I have nowhere near as much knowledge as Mark/Samah. But isn't putting cleanup of allocations etc in the finally block the way to deal with the 'if a random goto' has occurred how do I know whats run and what has not? (with tests on objects in the finally block to see if they have been instantiated etc, prior to being cleaned up, yes I know monkey has GC but I'm thinking of a contrived example)<br><br>That is you defer anything at risk of needing to be resolved regardless of exception to the finally stanza surrounded by checking code to make sure the cleanup is required?<br><br>Maybe Mark, you mean its undertaking that mental sorting of what lives in the try/catch block and what is deferred to final cleanup is what you mean by 'written very carefully'<br><br>Fascinating exchange anyway, made me think much more about the value versus complexity that exceptions resolve or introduce.<br><br>Roger. <br><br></td></tr></table><br>
<a name="2011372"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td class="posthead"><table width=100% cellspacing=0 cellpadding=0><tr ><td >Samah</td><td align="right"><font class=tiny>(Posted 2011)</font>&nbsp;<a href="#7">[#7]</a></td></tr></table></td></tr><tr ><td class="posttext"> <div class="quote"> But isn't putting cleanup of allocations etc in the finally block the way to deal with the 'if a random goto' has occurred how do I know whats run and what has not? <br></div><br>Yep, but I'm not sure that all targets support a finally keyword; I didn't look that far into it.  Regardless, as Mark said, it requires that the developer knows what he/she is doing.  :) <br><br></td></tr></table><br>
<a name="bottom"></a>
<table width=100% cellspacing=0 cellpadding=0><tr ><td align="right"><a href="http://monkeycoder.co.nz" target="_blank"><img src="/img/monkey2.svg" ></a> <a href="https://github.com/blitz-research" target="_blank"><img src="/img/github.svg" ></a> <a href="https://discord.gg/qJccbSp" target="_blank"><img src="/img/discord.svg" ></a></td></tr></table><br></div><br><table width="100%"><tr><td></body></html lang="en">
